### ISSUE TITLE
Phase 1: Bundling Design & Spec

### ISSUE THREAD

**Issue by ktinubu** [tinaudio/synth-setter#543]
## Overview

Capture the Surge XT bundling plan in a design doc and resolve the open questions that block implementation.

Part of #542.

## Scope

- Write `docs/design/bundling.md` covering goals, non-goals, package restructure, build pipeline, runtime plugin discovery, licensing, and open questions
- Investigate whether the prebuilt `Surge XT.vst3` bundle embeds factory wavetables/patches or needs a sibling `surge-data/` directory
- Confirm Linux glibc target for the prebuilt `.so` via `objdump -p | grep GLIBC`
- Confirm wheel size estimate by measuring the prebuilt Surge XT release asset

## Acceptance criteria

- [ ] `docs/design/bundling.md` merged
- [ ] Open question on factory-data layout has an answer (either "bundle is self-contained" or "bundle `surge-data/` and set `SURGE_DATA_PATH`")
- [ ] Linux glibc target chosen (likely `manylinux_2_28`)
- [ ] Wheel size estimate documented in the design doc

## Tasks

Sub-issues will be created as work is scoped.

**Comment by ktinubu** [comment]
# Design Doc: Bundling Surge XT into synth-setter

> **Status**: Draft
> **Author**: ktinubu@
> **Last Updated**: 2026-04-12
> **Tracking**: TBD

______________________________________________________________________

### Index

| §   | Section                                                                       | What it covers                                                      |
| --- | ----------------------------------------------------------------------------- | ------------------------------------------------------------------- |
| 1   | [Context & Motivation](#1-context--motivation)                                | Problem statement, user story                                       |
| 2   | [Typical Workflow](#2-typical-workflow)                                       | End-to-end install + import example                                 |
| 3   | [Goals, Non-Goals & Design Principles](#3-goals-non-goals--design-principles) | Scope, requirements, anti-goals                                     |
| 4   | [System Overview](#4-system-overview)                                         | How wheels ship prebuilt native binaries, package layout            |
| 5   | [Package Restructure](#5-package-restructure)                                 | `src/` and `pipeline/` → `synth_setter/`, bundled asset layout      |
| 6   | [Build Pipeline](#6-build-pipeline)                                           | GitHub Actions matrix, platform tagging, smoke tests                |
| 7   | [Runtime Plugin Discovery](#7-runtime-plugin-discovery)                       | `importlib.resources` lookup, Hydra defaults, test migration        |
| 8   | [Design Decisions](#8-design-decisions)                                       | `pyproject.toml` changes, licensing, Dockerfile update              |
| 9   | [Open Questions & Risks](#9-open-questions--risks)                            | Known gaps and trade-offs                                           |
| 10  | [Implementation Sequence](#10-implementation-sequence)                        | PR-sized chunks from rename through contributor docs                |
| 11  | [Related Issues](#11-related-issues)                                          | Links into the existing issue graph                                 |

______________________________________________________________________

## 1. Context & Motivation

Topline goal: `pip install <wheel URL>` gives users the full synth-setter pipeline (data generation, training, evaluation, inference) with the Surge XT VST3 binary and factory presets bundled into the Python wheel. No manual plugin install, no CMake, no sudo. Everything installs into a local Python venv.

Today, getting synth-setter running end-to-end requires installing Surge XT separately: downloading a DAW plugin installer (macOS), building from source with CMake (Linux), or relying on a hardcoded `/usr/lib/vst3/Surge XT.vst3` path (test suite). This is a consistent friction point for new researchers and collaborators, and it blocks the evaluation CLI ([#491](https://github.com/tinaudio/synth-setter/issues/491)) from being usable without a full dev environment.

### User story

A researcher or musician creates a fresh virtualenv, runs `pip install <wheel URL>`, imports `synth_setter`, points it at audio data, and trains or runs inference without ever touching a DAW plugin installer, CMake, or a shell script. On Linux they are expected to have `xvfb` installed as a system package; a clear error points them there if it is missing.

______________________________________________________________________

## 2. Typical Workflow

```bash
# 1. Create a fresh virtualenv
python -m venv .venv
source .venv/bin/activate

# 2. Install the prebuilt wheel for the current platform
pip install https://github.com/tinaudio/synth-setter/releases/download/v<x>/synth_setter-<x>-<platform>.whl

# 3. Import and use — Surge XT is bundled, no manual plugin install
python -c "from synth_setter.engine import load_plugin; load_plugin()"
```

Wheel naming at launch:

```text
synth_setter-<ver>-py3-none-manylinux_2_28_x86_64.whl
synth_setter-<ver>-py3-none-macosx_11_0_arm64.whl
```

______________________________________________________________________

## 3. Goals, Non-Goals & Design Principles

### Goals

- **One-command install.** `pip install <wheel URL>` gives the user the full pipeline — ML code, distributed data pipeline, bundled Surge XT VST3, bundled factory presets, LICENSE/NOTICE — in a local virtualenv with no sudo, no CMake, and no manual plugin install.
- **Platform-tagged wheels.** Each wheel targets one OS/arch pair (macOS arm64, Linux x64) and is smoke-tested on that exact platform in CI before release.
- **Library-first.** synth-setter is importable as `synth_setter` from any Python project. A user-facing CLI is deferred — the wheel enables library use today.
- **Licensing cleanliness.** Surge XT is GPL-3.0 and synth-setter is GPL-3.0 — the wheel ships the upstream LICENSE and a NOTICE file pointing at Surge XT source per GPL-3.0 requirements.
- **Headless-compatible.** On Linux, `xvfb` is a documented system prerequisite and the engine surfaces a clear error if it is missing.

### Design Principles

- **Prebuilt native binaries follow the torch/onnxruntime pattern.** Wheels are platform-tagged zip archives; pip picks the right one. No compilers required on the user's machine.
- **No C extensions.** synth-setter is pure Python — platform tagging is a `bdist_wheel` subclass trick, not a C build.
- **`importlib.resources` for asset lookup.** Bundled paths resolve through the installed package, not relative paths or hardcoded system locations.
- **Manual build matrix, not cibuildwheel.** There is no C extension to compile, so cibuildwheel's manylinux container logic adds complexity without benefit.
- **PR-sized chunks.** The rename and restructure happens incrementally, not in one mega-commit — imports, Hydra YAML defaults (`plugin_path`), scripts, Dockerfile, and tests all move together per PR.

### What This System Deliberately Avoids

- **Windows, Linux arm64, macOS x64.** Out of scope for v1.
- **PyPI publishing.** GitHub Releases only until end-to-end flow is proven.
- **Replacing RunPod for production-scale data generation.** Local data gen from the wheel is for dev/test; large runs still go through the distributed pipeline.
- **A user-facing CLI (`synth-setter <cmd>`).** Library-only in v1; CLI shape is deferred to a follow-up design doc.
- **Bundled default model weights.** Weights are downloaded on first use; the source (HF Hub, W&B, GH Release asset) is TBD.
- **Solving Linux headless plugin hosting.** Users provide their own Xvfb via the system package manager.

### Success Metrics

| Metric                     | Target                                                                | How to Measure                                                 |
| -------------------------- | --------------------------------------------------------------------- | -------------------------------------------------------------- |
| One-command install        | `pip install <wheel URL>` succeeds in a clean venv                    | CI smoke test on macOS arm64 and Linux x64                     |
| Plugin loads out of the box| `load_plugin()` succeeds without any user-provided plugin path        | CI smoke test inside the same clean venv                       |
| Platform coverage          | Wheels built and released for macOS arm64 and Linux x64               | GitHub Release assets                                          |
| Licensing compliance       | Wheel ships Surge XT LICENSE + NOTICE + source link                   | Inspect wheel contents; README audit                           |
| Wheel size                 | Per-platform wheel fits comfortably in GitHub Release asset limits    | Measure actual size after first build (open question §9.5)     |

### Non-Goals

- **Multi-platform parity beyond v1.** Windows, Linux arm64, and macOS x64 may follow if demand justifies them, but they are not required to ship v1.
- **Publishing to PyPI.** GitHub Releases is sufficient until the end-to-end install flow is proven.
- **Replacing the distributed data pipeline.** The bundled wheel enables local generation for dev and small experiments; research-scale runs still go through RunPod per [`data-pipeline.md`](data-pipeline.md).
- **A CLI surface.** `synth-setter <cmd>` is explicitly deferred — see open question §9.6.

______________________________________________________________________

## 4. System Overview

Python wheels are zip archives with a platform tag encoded in the filename. pip reads the tag and downloads the matching one for the user's OS and arch. Projects like torch, onnxruntime, and tensorflow use this to ship precompiled native binaries without requiring users to install compilers. Installation goes into the user's virtualenv — no sudo.

Since synth-setter is pure Python, the platform tag is forced via a `bdist_wheel` subclass (setuptools trick). There is no C extension to compile.

### Scope (v1)

| Concern               | v1 choice                                                                            |
| --------------------- | ------------------------------------------------------------------------------------ |
| **Platforms**         | macOS arm64, Linux x64 (manylinux)                                                   |
| **Hosting**           | GitHub Releases only                                                                 |
| **Package contents** | `synth_setter.*` (ML code from today's `src/`), `synth_setter.pipeline.*` (distributed pipeline from today's `pipeline/`), bundled `Surge XT.vst3`, bundled `.vstpreset` files, bundled LICENSE/NOTICE |
| **Model weights**     | Downloaded on first use (source TBD — HF Hub, W&B, or GH Release asset)              |
| **Entry point**       | Library only                                                                         |

______________________________________________________________________

## 5. Package Restructure

Rename `synth-permutations` to `synth-setter`. Move today's top-level trees into a single installable package:

```text
Before                          After
------                          -----
src/                            synth_setter/
pipeline/                       synth_setter/pipeline/
plugins/Surge XT.vst3/          synth_setter/bundled/Surge XT.vst3/
presets/*.vstpreset             synth_setter/bundled/presets/
                                synth_setter/bundled/LICENSES/surge-xt.txt
                                synth_setter/bundled/NOTICE
```

`python -m pipeline` becomes `python -m synth_setter.pipeline`. This is a large rename touching imports, Hydra YAML defaults (`plugin_path`), scripts, the Dockerfile, tests, and the generate entrypoint subprocess invocation. It happens in PR-sized chunks, not one mega-commit.

______________________________________________________________________

## 6. Build Pipeline

GitHub Actions runs a platform matrix, assembles the `bundled/` tree from upstream Surge XT releases, forces the platform tag, smoke-tests the wheel, and uploads it as a release asset.

```text
matrix: [macos-14 (arm64), ubuntu-22.04 (x64)]
  ↓
download prebuilt Surge XT release asset for target platform from
  https://github.com/surge-synthesizer/releases-xt/releases
  ↓
extract Surge XT.vst3 → synth_setter/bundled/
copy presets/*.vstpreset → synth_setter/bundled/presets/
copy Surge XT LICENSE / NOTICE → synth_setter/bundled/LICENSES/
  ↓
python -m build --wheel
  ↓
force platform tag via bdist_wheel override
  (plat_name = manylinux_2_28_x86_64 or macosx_11_0_arm64)
  ↓
smoke-test the wheel in a clean venv on the target platform:
  python -c "from synth_setter.engine import load_plugin; load_plugin()"
  ↓
upload wheel as GitHub Release asset
```

Manual matrix, not cibuildwheel — there is no C extension to compile, so cibuildwheel's manylinux container logic adds complexity without benefit.

______________________________________________________________________

## 7. Runtime Plugin Discovery

Plugin and preset paths resolve via `importlib.resources` from the installed package, replacing today's relative-path defaults (`plugins/Surge XT.vst3`) and hardcoded test paths (`/usr/lib/vst3/Surge XT.vst3`):

```python
from importlib.resources import files

def get_plugin_path() -> str:
    return str(files("synth_setter.bundled") / "Surge XT.vst3")

def get_preset_path(name: str) -> str:
    return str(files("synth_setter.bundled.presets") / f"{name}.vstpreset")
```

Callers in `src/data/vst/core.py` and `src/data/vst/generate_vst_dataset.py` switch from CLI-arg defaults to this lookup. Hydra YAML dataset configs keep `plugin_path` as an overridable field whose default resolves to the bundled path.

______________________________________________________________________

## 8. Design Decisions

### 8.1 `pyproject.toml` changes

- `name = "synth-setter"` (was `synth-permutations`)
- `[tool.setuptools.package-data]` includes `synth_setter = ["bundled/**/*"]`
- `bdist_wheel` subclass that forces `plat_name` for platform tagging
- Existing dependencies unchanged: pedalboard, pytorch-lightning, hydra-core, h5py, pyloudnorm, wandb

### 8.2 Testing strategy

- **Build-time smoke.** CI installs each wheel into a clean venv on its target platform and loads the plugin.
- **`requires_vst` test migration.** Update existing tests to resolve the plugin via `importlib.resources` instead of `/usr/lib/vst3/Surge XT.vst3`.
- **Dockerfile update.** Drop the Surge XT download/build stages; the wheel bundles the plugin. Keep the Xvfb wrapping.

### 8.3 Licensing

Surge XT is GPL-3.0 and synth-setter is GPL-3.0 — compatible. The wheel ships:

- `synth_setter/bundled/LICENSES/surge-xt.txt` — Surge XT LICENSE
- `synth_setter/bundled/NOTICE` — attribution and link to Surge XT source
- README link to Surge XT source per GPL-3.0 requirements

______________________________________________________________________

## 9. Open Questions & Risks

| #   | Question / Risk                                                                                                                                                | Impact                                    | Status |
| --- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------- | ------ |
| 1   | Does the prebuilt `Surge XT.vst3` bundle embed its factory wavetables and patches, or does it expect a sibling `surge-data/` directory?                       | Blocks macOS-vs-Linux binary layout       | Open   |
| 2   | Surge XT version strategy — pin a specific release and bump manually, or run a CI cron that auto-rebuilds wheels on new Surge releases?                       | Maintenance cadence                       | Open   |
| 3   | Should `SYNTH_SETTER_PLUGIN_PATH` env var let users point at their own Surge XT install (useful for dev, version pinning)?                                     | Dev ergonomics                            | Open   |
| 4   | Linux glibc target — check the prebuilt `.so` with `objdump -p \| grep GLIBC` to pick the right manylinux tag (likely `manylinux_2_28`)                       | Platform tag correctness                  | Open   |
| 5   | Wheel size — measure actual size after the first build. If per-platform wheels exceed a few hundred MB, reconsider whether any assets should be split out.    | Release asset size, download UX           | Open   |
| 6   | User-facing CLI shape — what `synth-setter <cmd>` subcommands should exist, if any. Library-only in v1, deferred to a follow-up design doc.                   | Future UX                                 | Open   |

Open question 1 is blocking: if a separate `surge-data/` directory is required, it must be bundled into `synth_setter/bundled/` and `SURGE_DATA_PATH` set before `pedalboard.load_plugin()`. This is investigated during step 4 of §10.

Open questions 2 and 3 are deferred until after the first wheel works end-to-end.

______________________________________________________________________

## 10. Implementation Sequence

Each step is a PR-sized chunk:

1. **Rename package.** `synth-permutations` → `synth-setter` in `pyproject.toml` and all references. No bundling yet. Closes [#134](https://github.com/tinaudio/synth-setter/issues/134).
2. **Move `src/` → `synth_setter/`.** Grep all file types (`.py`, `.yaml`, `.md`, `.toml`, `.sh`, `Dockerfile`) for old paths. Related to [#301](https://github.com/tinaudio/synth-setter/issues/301).
3. **Move `pipeline/` → `synth_setter/pipeline/`.** Update `python -m pipeline` to `python -m synth_setter.pipeline`; update the generate entrypoint subprocess call.
4. **Bundle Surge XT + presets.** Add files under `synth_setter/bundled/`, switch lookup to `importlib.resources`, update `requires_vst` tests. Investigate open question §9.1 (surge-data) at this step. Closes [#487](https://github.com/tinaudio/synth-setter/issues/487).
5. **Wheel build script.** `scripts/build_wheel.sh` that downloads the prebuilt Surge XT release, assembles `bundled/`, and runs `python -m build`.
6. **GitHub Actions workflow.** Matrix build (macos-14, ubuntu-22.04), smoke test, release asset upload.
7. **Contributor docs.** README install instructions, Linux Xvfb note, licensing and NOTICE.

______________________________________________________________________

## 11. Related Issues

| Issue | Title                                                                                               | Relationship                                                                                                                                                                               |
| ----- | --------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| [#134](https://github.com/tinaudio/synth-setter/issues/134) | `fix(build): pyproject.toml package name still synth-permutations`                                 | Closed by step 1                                                                                                                                                                           |
| [#301](https://github.com/tinaudio/synth-setter/issues/301) | `refactor(code-health): rename src/ package to avoid mutmut and src-layout conflicts`              | Related to step 2                                                                                                                                                                          |
| [#487](https://github.com/tinaudio/synth-setter/issues/487) | `bug(testing): Surge XT plugin path is hardcoded — cross-platform discovery needed`                | Closed by step 4 — bundling replaces the proposed env-var-and-fallback solution; the env var override survives as open question §9.3                                                      |
| [#491](https://github.com/tinaudio/synth-setter/issues/491) | `feat(evaluation): synth matching CLI — predict params from audio and apply to Surge XT`           | Future consumer that becomes much simpler once the bundled wheel ships                                                                                                                     |

______________________________________________________________________
